### PR TITLE
Some improvement or fixes

### PR DESCRIPTION
**Pull request by raoyitao**
* Fix an issue that when testcase has a timeout, the exception info
  cannot be caught, and our code itself raises an `IndexError`.
* Enhancement for `HttpExporter` that if the report has some objects
  that cannot be serialized by json, then we provide a customized
  json encoder to deal with it to avoid exception.
* Customized name_func for parametrized testcase cannot ensure no
  deplicated testcase names exist: after we simply renaming the invalid
  names there might still conflict be found, we need to check again.

### CODE REVIEW

File: doc/en/multitest.rst [hunk 1]
```diff
@@ -1024,11 +1026,18 @@ for sample implementation.
       """
       number_list = kwargs['number_list']
 
-      return '{func_name}__numbers_{joined_list}__result_{expected}'.format(
-          func_name=func_name,
-          joined_list='_'.join(map(str, number_list)),
-          expected=kwargs['expected']
-      )
+      if kwargs.get('for_display', False):
```
Comment (ryan-collingham) [inline doc/en/multitest.rst]: I think this is over-complicating. Surely the custom name generator can just return a single human-readable name? Why does it need to be a valid python identifier?

File: testplan/common/utils/timing.py [hunk 2]
```diff
@@ -92,10 +95,14 @@ def timeout(seconds, err_msg='Timeout after {} seconds.'):
     def timeout_decorator(func):
         """"""
         def _new_func(result, old_func, old_func_args, old_func_kwargs):
-            result.append(old_func(*old_func_args, **old_func_kwargs))
+            try:
+                result.append(old_func(*old_func_args, **old_func_kwargs))
+            except Exception as exp:
+                result[0] = False
+                result.append(format_trace(inspect.trace(), exp))
```
Comment (ryan-collingham) [inline testplan/common/utils/timing.py]: It's not very important but I think that `traceback.format_exc()` would be simpler here, I don't see why we need a custom function for formatting stack traces when it's right there in the standard library.

File: testplan/exporters/testing/http/__init__.py [hunk 3]
```diff
@@ -14,6 +16,12 @@
 from ..base import Exporter
 
 
+class CustomJsonEncoder(json.JSONEncoder):
```
Comment (ryan-collingham) [inline testplan/exporters/testing/http/__init__.py]: Why is this needed exactly?

File: testplan/report/testing/base.py [hunk 4]
```diff
@@ -552,6 +552,7 @@ class TestGroupReport(BaseReportGroup):
 
     def __init__(self,
                  name,
+                 display_name=None,
```
Comment (ryan-collingham) [inline testplan/report/testing/base.py]: We definitely don't need an additional display_name parameter on the report. If a custom name is specified for a testcase/testsuite, that becomes the `name` parameter. Only if no name is specified do we use the function name as a default value. The report object only ever has one name.

File: testplan/report/testing/schemas.py [hunk 5]
```diff
@@ -195,11 +195,15 @@ class TestCaseReportSchema(ReportSchema):
...
     status_override = fields.String(allow_none=True)
 
+    display_name = fields.String(allow_none=True)
+
     entries = fields.List(EntriesField())
 
     status = fields.String(dump_only=True)
     runtime_status = fields.String(dump_only=True)
     counter = fields.Dict(dump_only=True)
+    passed = fields.Bool()
```
Comment (ryan-collingham) [inline testplan/report/testing/schemas.py]: We don't need these extra fields - we already send the status, this information is redundant.

File: testplan/testing/multitest/suite.py [hunk 6]
```diff
@@ -382,12 +383,13 @@ def wrapper(function):
         function.__tags__ = copy.deepcopy(tag_dict)
 
         if parameters is not None:  # Empty tuple / dict checks happen later
-
+            function.__display_name__ = name
```
Comment (ryan-collingham) [inline testplan/testing/multitest/suite.py]: Here, you should set `function.__display_name__ = name if name else function.__name__`.

Ideally this wouldn't be a dunder attribute (i.e. just `function._display_name`, however since that is the established convention it's probably best to stick with it for now.

File: testplan/testing/multitest/base.py [hunk 7]
```diff
@@ -340,6 +340,7 @@ def dry_run(self, status=None):
                 testcase_report = TestCaseReport(
                     name=testcase.__name__,
```
Comment (ryan-collingham) [inline testplan/testing/multitest/base.py]: Here and below you should set `name=testcase.__display_name__` and not set any `display_name` parameter in the report.

File: doc/en/multitest.rst (hunk 1 above)
Comment (raoyitao) [reply doc/en/multitest.rst]: the name_func is used for generating names for parametrized testcases. Testplan will check if the generated name is not a valid python identifier then a warning message is printed, and testplan will simply reset the name with suffix. So user can know that their customized naming function does not work well. So, this name_func is designed to generate function name, not human readable display name. I also think that this is over-complicating, it is better to use another default name_func for generating hunan-readable names.

The question is: should we provide another another argument to let user customize the name_func for display name? if we should, that seems a little complicated, if we do not, then an unwanted name can be displayed such as "My test function (arg1: 1, arg2: [{'a': 'b'}, {'c': 'd'} ...])"

File: testplan/common/utils/timing.py (hunk 2 above)
Comment (raoyitao) [reply testplan/common/utils/timing.py]: I' m not sure, I saw that the __exit__() function of Exception Logger calls this format_trace. Is it some different ?

File: testplan/report/testing/base.py (hunk 4 above)
Comment (raoyitao) [reply testplan/report/testing/base.py]: Yes the display name should be the name of testcase here, my concern is , if we totally remove that original function name from report, then it is hard for us to debug or trace with that json report, or maybe one day use want a feature that switch back to testcase identifier on UI. It makes sense that the name of report is 'display name', just like that name of Multitest is a human readable string, but somehow I think we should also keep an attribute 'testcase' or 'testcase_name' in the report.

File: testplan/report/testing/schemas.py (hunk 5 above)
Comment (raoyitao) [reply testplan/report/testing/schemas.py]: because in another PR you left a comment that don't check the pass or fail status for a testcase, it already have the attribute. There the report has already been dumped to json so no 'passed' or 'failed' key is kept. But I agree to remove the attributes here, because in a recent commit, it add an unstable status, for example, a skipped testcase is unstable, but if a suite has 2 cases: passed and unstable, then the suite is passed, so RFB cannot reply on 'passed' or 'failed' attributes because in RFB check it has different meaning, finally it is either pass or fail, even all testcases are set skipped, it passes but no unstable status should be reported. So I think I should check the status again in the rfb exporter.

File: testplan/testing/multitest/suite.py (hunk 6 above)
Comment (raoyitao) [reply testplan/testing/multitest/suite.py]: make sense

File: doc/en/multitest.rst [hunk 8]
```diff
@@ -1011,8 +1011,10 @@ However you can provide custom name generation functions to override this
 functionality via ``@testcase(parameters=..., name_func=custom_name_func)``
 syntax. You just need to implement a function that accepts
 ``func_name`` and ``kwargs`` as arguments, ``func_name`` being a string and
-``kwargs`` being an ``OrderedDict``.
-See
+``kwargs`` being an ``OrderedDict``. You may check if there is an argument
+``for_display`` in `kwargs``, if the value is true, then it is suggested to
```
Comment (Pyifan) [inline doc/en/multitest.rst]: I agree with Ryan, I don't think we need another for_display here. If this is the display name we discussed, it does not belong here (parameterization session).

File: doc/en/multitest.rst (hunk 1 above)
Comment (Pyifan) [reply doc/en/multitest.rst]: Does not seem necessary to me at least for now. The idea is if user specify his own name_func, then he takes full control. I do not think it is super difficult to come up with a pattern that is both human-readable && python identifier valid.

File: testplan/testing/multitest/parametrization.py [hunk 9]
```diff
@@ -325,14 +347,21 @@ def default_name_func(func_name, kwargs):
...
+    else:
+        arg_strings = ['{arg}_{{{arg}}}'.format(arg=arg) for arg in kwargs]
+        template = '{func_name}__' + '__'.join(arg_strings)
     return template.format(func_name=func_name, **kwargs)
 
 
 def generate_functions(
     function,
     parameters,
+    name,
```
Comment (Pyifan) [inline testplan/testing/multitest/parametrization.py]: doc string?

File: testplan/testing/multitest/suite.py [hunk 10]
```diff
@@ -357,6 +357,7 @@ def _testcase(func):
 
 
 def _testcase_meta(
+    name=None,
```
Comment (Pyifan) [inline testplan/testing/multitest/suite.py]: I think it's confusing to have a name param in @testcase decorator. Intuitively the name of a testcase is the function name. We can have a display_name for the testcase though. I suggest we change name ==> display_name here.

File: testplan/report/testing/base.py (hunk 4 above)
Comment (Pyifan) [reply testplan/report/testing/base.py]: I prefer that we keep the original testcase name (func name) unchanged in the report, just add one additional field say display_name. Changing the name itself might have other side effect that we cannot predict right now.

**Comment by Pyifan** [comment]
I think adding 'display name' for testcase is a very good addition! I prefer that we add this as an additional field in the report, and leave it to the Exporter to render the user friendly name. On the other hand we need to caution with the complexity when 'display name' encounters 'parameterization'. My suggestion is to redesign @testcase decorator as:

@testcase(display_name=...) for single testcase
@testcase(display_name_func=...) for parameterized testcase
and get rid of the original
@testcase(name_func=...)

File: testplan/exporters/testing/http/__init__.py (hunk 3 above)
Comment (raoyitao) [reply testplan/exporters/testing/http/__init__.py]: user can define attributes in test suites, but we don't have a check that all their attributes can be serialized by JSON (this rarely happens, but user can define weird things like some objects of non-basic type). To avoid exception in dumping JSON so we can add this customized encoder. Xml pdf exported is not affected because they choose the pre-defined attributes to dump. I think Json report is the most basic if we want to debug or trace, so it should contains all the information. Json exporter also need it so I move this Encoder to exporters/testing/json/__init__.py and HTTPExporter imports it.

File: doc/en/multitest.rst (hunk 1 above)
Comment (ryan-collingham) [reply doc/en/multitest.rst]: Why does the name need to be a valid python identifier?

I agree it's probably not needed for now.

File: testplan/common/utils/timing.py (hunk 2 above)
Comment (ryan-collingham) [reply testplan/common/utils/timing.py]: I don't think we need it, it just does the same as standard library function.

File: testplan/exporters/testing/http/__init__.py (hunk 3 above)
Comment (ryan-collingham) [reply testplan/exporters/testing/http/__init__.py]: ok, I think that if a user defines some weird attribute on a testsuite that is quite unusual and we shouldn't worry about making it work. But if it's only a simple change it's fine to leave in.

File: testplan/report/testing/base.py (hunk 4 above)
Comment (ryan-collingham) [reply testplan/report/testing/base.py]: No, as discussed absolutely we cannot have more than one name in the report as it will over complicate the UI code too much. One name only please.

File: testplan/report/testing/schemas.py (hunk 5 above)
Comment (ryan-collingham) [reply testplan/report/testing/schemas.py]: I think you are confusing with something else. Please can you just remove these passed/failed fields, the JSON should not contain redundant data to minimise its size.

File: testplan/testing/multitest/suite.py (hunk 10 above)
Comment (ryan-collingham) [reply testplan/testing/multitest/suite.py]: the whole point of this PR is to provide a way to override the name of a testcase via the decorator. The function name is just the default string used for the name, it has no other special meaning.

File: doc/en/multitest.rst (hunk 1 above)
Comment (raoyitao) [reply doc/en/multitest.rst]: we have decided to add display name, so the name_func is now used for generating display name for parametrized testcases. and our default name func should be improved a little, because most time the string format of parameter is not simple positive integer or letters, and they will be renamed again.

File: testplan/common/utils/timing.py (hunk 2 above)
Comment (raoyitao) [reply testplan/common/utils/timing.py]: I will check it and if not necessary I will remove it.

File: testplan/exporters/testing/http/__init__.py (hunk 3 above)
Comment (raoyitao) [reply testplan/exporters/testing/http/__init__.py]: It's only a few lines change, more safe, some team like proxima like to dynamically throw things into modules or classes. Let me keep it.

File: testplan/report/testing/schemas.py (hunk 5 above)
Comment (raoyitao) [reply testplan/report/testing/schemas.py]: I just checked with Yuxuan/Yifan about the changes made for Report status. I moved the attributes.

File: testplan/report/testing/base.py (hunk 4 above)
Comment (raoyitao) [reply testplan/report/testing/base.py]: I think we have different point of views.
**Ryan**: only a `name` attribute added into report, inside testcase it's actually __display_name__, no redundant attributes kept in report.
**Yifan**: `name` is intuitively the defined name of testcase, so a `display_name` is added in both testcase decorator and test report.
**Me**: `name` is added for testcase decorator because it's short and regular, but in json report we keep both original name and display name, when users search on UI, I think both of them are needed, just like when you search a people, the real name and user id will both be considered.

Comment (ryan-collingham) [reply testplan/report/testing/base.py]: I don't really care too much about what attributes you set on the testcase functions but, to reiterate, we must not overcomplicate the report by adding an extra "display_name" field. The name parameter of testcase is an override, that becomes the testcases unique identifier and is the only name used for search and filter.

File: testplan/common/utils/timing.py (hunk 2 above)
Comment (raoyitao) [reply testplan/common/utils/timing.py]: Hi Ryan, as I check again, **format_trace(inspect.trace(), exc))** is used at many places in open source repo. To avoid making it not stable so I won't change it in this pull request. If you though it's necessary then another PR can be opened.

Comment (ryan-collingham) [reply testplan/common/utils/timing.py]: that's fine, I wasn't suggesting you remove it in this PR. I was more just asking the broader question of do we need it, and do you need to use it right here. We can look at removing in another PR, if it can be replaced by the traceback module.
